catatonit.c

```diff
@@ -136,6 +144,93 @@ int kernel_signals[] = {SIGFPE, SIGILL, SIGSEGV, SIGBUS, SIGABRT, SIGTRAP, SIGSY
 
 #define ARRAY_LEN(arr) (sizeof(arr) / sizeof(*arr))
 
+#ifndef HAVE_CLOSE_RANGE
+static int close_range(unsigned int fd, unsigned int max_fd, unsigned int flags)
+{
+# ifdef __NR_close_range
+  return (int) syscall(__NR_close_range, fd, max_fd, flags);
+# else
+  errno = ENOSYS;
+  return -1;
+# endif
+}
+#endif
+
+/*
+ * Close every fd >= n that is different from exclude_fd using close_range.
+ */
+static int close_range_fds_ge_than(int n, int exclude_fd)
+{
+	int r, saved_errno = 0;
+
+	/* exclude_fd is not in the [n, UINT_MAX] range.  */
+	if (exclude_fd < n)
+		return close_range(n, UINT_MAX, 0);
+
+	/* exclude_fd is the first fd in the [n, UINT_MAX] range.  */
+	if (exclude_fd == n)
+		return close_range(n + 1, UINT_MAX, 0);
+
+	/* exclude_fd is between n and UINT_MAX.  */
+	errno = 0;
+	r = close_range(n, exclude_fd - 1, 0);
+	/*
+	 * attempt to close as many FDs as possible but return an error
+	 * if the close_range() failed.
+	 */
+	if (exclude_fd < UINT_MAX) {
+		saved_errno = errno;
+		r = close_range(exclude_fd + 1, UINT_MAX, 0);
+		/* If the previous call failed, restore errno.  */
+		if (saved_errno != 0) {
+			r = -1;
+			errno = saved_errno;
+		}
+	}
+	return r;
+}
+
+/*
+ * Close every fd >= n that is different from exclude_fd.
+ */
+static void close_fds_ge_than(int n, int exclude_fd)
+{
+	struct dirent *next;
+	DIR *dir;
+	int fd;
+
+	if (close_range_fds_ge_than(n, exclude_fd) == 0)
+		return;
+
+	/* Fallback when close_range fails.  */
+	debug("close_range() failed, fallback to close() each open FD: %m");
+
+	dir = opendir("/proc/self/fd");
+	if (dir == NULL) {
+		warn("cannot opendir /proc/self/fd: %m");
+		return;
+	}
+
+	fd = dirfd(dir);
+	for (next = readdir(dir); next; next = readdir(dir)) {
+		long long val;
+		const char *name = next->d_name;
+		int r;
+
+		if (name[0] == '.')
+			continue;
+
+		val = strtoll(name, NULL, 10);
+		if (val < n || val == fd || val == exclude_fd)
+			continue;
+
+		r = close(val);
+		if (r < 0)
+			warn("cannot close %d: %m", val);
```

Comment:
If we add a warning when calling this function, maybe just make this a debug message? It'd probably be better to be a bit less chatty -- a single warning saying "failed to close some file descriptor in range >=3" is fine.

Reply:
changed to debug message and added just a warning for `close_fds_ge_than`